### PR TITLE
[MONO][WASM] Fix System.ComponentModel.TypeConverter tests by  adding missing ILLink dependencies

### PR DESCRIPTION
**Pull request by naricc**
When EnableAggressiveTrimming=true, these types get trimmed out, which causes the  System.ComponentModel.TypeConverter to fail. This adds the missing assemblies/types to ILLink.Descriptor.xunit.xml, preserving them during the trimming process, and also introduces a project specific ILLink descriptor.

### CODE REVIEW

**Comment by mdh1418** [comment]
Is it known why parts of `Castle.Core` are getting linked out? Does it look like a linker bug or is it some reflection related reason that the linker will never be able to detect the assembly is needed?

For either reason, I think it is worth noting either in this PR or possibly as a comment above the assembly inclusion in the `ILLink.Description.xunit.xml` file.

On another note, maybe it is worth renaming `ILLink.Description.xunit.xml` since it will not be xunit specific assemblies. Or maybe we can add another `.xml` file that will include test suite needed assemblies (that are not xunit) and add that to https://github.com/dotnet/runtime/blob/8445e4eaa3b18bfdbed1eb1507293777c590d371/eng/testing/tests.mobile.targets#L242

```
    <ItemGroup>
      <TrimmerRootDescriptor Include="$(MSBuildThisFileDirectory)ILLink.Descriptor.xunit.xml" />
      <TrimmerRootDescriptor Include="$(MSBuildThisFileDirectory)ILLink.Descriptor.<some-name>.xml" />
    </ItemGroup>
```

**Comment by naricc** [comment]
@mdh1418 

Adding another ILLink file makes sense; should I do it in this PR, or a follow up?

The actual place that ends up needing theses constructors is here (and similar things): https://github.com/castleproject/Core/blob/d88e94447150996973d947a150ffda5efc0c3c1c/src/Castle.Core/DynamicProxy/Tokens/InvocationMethods.cs#L28

This is using reflection, but it guess it is conceivable that some sufficiently smart static analysis could figure that out. How clever is the linker expected to be?

**Comment by mdh1418** [comment]
@naricc I think adding it to this PR would make sense in that in addition to adding `xunit.runner.utility.netcoreapp10` in `ILLink.Descriptor.xunit.xml` to get `System.ComponentModel.TypeConverter` tests to pass, non xunit assemblies need to be linked in as well, by being introduced in a file alongside `ILLink.Descriptor.xunit.xml`.

But it seems like `Castle.Core` cannot be resolved on the `runtime` pipeline `Build Browser wasm Release AllSubsets_Mono` lane

> How clever is the linker expected to be?

I don't know in detail, but I understand there is general issues with reflection. @marek-safar may know better for this situation.

**Comment by vitek-karas** [comment]
The sample you noted above https://github.com/castleproject/Core/blob/d88e94447150996973d947a150ffda5efc0c3c1c/src/Castle.Core/DynamicProxy/Tokens/InvocationMethods.cs#L28 is something linker should be able to figure out just fine. You can try setting `SuppressTrimAnalysisWarnings=false` and see all the warnings coming from the "castle code" - linker generates these warnings everytime it sees a pattern it can't handle.

**Comment by naricc** [comment]
@vitek-karas So I ran it with SuppressTrimAnalysisWarnings=false, and it seems like it is not actually lossing track of it where I thought it was. Here are all the castle related errors: https://gist.github.com/naricc/f3cb83c71bb804bce8548331b62a220c

There isn't one involving CompositionInvocation invocation constructor. But I am then unsure why including those assemblies in TrimmerRootDescriptor makes any difference?

**Comment by vitek-karas** [comment]
Found the issue - the problem is that `Castle.Core` uses special extension methods over reflection - they look identical in C# and they added them into `System.Reflection` namespace, so it's practically invisible in source file. But in IL they don't call `Type.GetConstructor`, instead they call `NetCoreReflectionExtensions.GetConstructor` which is not annotated.

Running with `SuppressTrimAnalysisWarnings=false` produces lot of warnings, one of which is:
```console
ILLink: Trim analysis error IL2067: System.Reflection.NetCoreReflectionExtensions.GetConstructor(Type,BindingFlags,Object,Type[],Object[]): 'type' argument does not satisfy 'DynamicallyAccessedMemberTypes.PublicConstructors', 'DynamicallyAccessedMemberTypes.NonPublicConstructors' in call to 'System.Reflection.TypeExtensions.GetConstructors(Type,BindingFlags)'. The parameter 'type' of method 'System.Reflection.NetCoreReflectionExtensions.GetConstructor(Type,BindingFlags,Object,Type[],Object[])' does not have matching annotations. The source value must declare at least the same requirements as those declared on the target location it is assigned to.
```

Which is the root of the problem. The fix in this case would be to annotate the method in Castle, but honestly, Castle has so many other warnings... not sure if this one alone would fix the larger problem.

**Review by radical — APPROVED** [review]
LGTM 👍 
In a follow up PR (wip in https://github.com/dotnet/runtime/pull/50885), this will get used in other tests also.

File: src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml
Comment (mdh1418) [inline src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml:1]: Nit: a comment at the top of this file to give visibility into what this file does/gets consumed  and why it is needed

File: src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml
Comment (mdh1418) [inline src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml:8]: Out of curiosity, are there specific parts of these assemblies that show up in `System.ComponentModel.TypeConverter` and need to not be linked out? Are `xunit` specific assemblies not only related to the test runner? (As in why is having these in `eng/testing/ILLink.Descriptor.xunit.xml` not sufficient?)

File: src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml
Comment (mdh1418) [inline src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml:9]: I'm curious if multiple test suites also require assemblies such as `mscorlib` and perhaps `System.PlatformDetection`, would we want to have a common `ILLink.Descriptor.xml` file the test suites can leverage. I guess that will depend on how many test suites end up needing these `ILLink.Descriptor.xml` files.

Comment (radical) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml:9]: The follow up PR will have many more descriptor files, specific to projects. We can make these changes there. Eg. https://github.com/dotnet/runtime/pull/50885/files#diff-d74ed78dff2160a861191b360b9a3a0b91a0fed05a1109728cc5472801a28799

The process I'm following is to add for specific project first, where it shows up. And then if it shows up in some other projects, then separate that out into a `eng/tests/ildescriptors/` directory, and use that in those projects. If any of these show up in *all* the projects, then we can add them to the "global" ones.

Comment (naricc) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml:9]: I think we should figure out what is common as I fix more tests suites, and move them into common files as necessary.

File: src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml
Comment (naricc) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml:8]: I think it is actually redundant. I will move them into ILLink.Descriptor.xunit.xml

File: src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml
Comment (naricc) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml:1]: Will do.

File: src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml
Comment (eerhardt) [inline src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml:1]: The name of the file should follow the linker naming pattern. There should be an `s` in `Descriptors`.

`ILLink.Descriptors.xml`

See the docs here:
https://github.com/mono/linker/blob/55bd0ebd00ddbf2cd817314b1cf6824003cfb463/docs/data-formats.md#descriptor-format

File: src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml
Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml:9]: Are these facade assemblies necessary with the latest linker bits?  I know changes went in to preserve facade assemblies that are necessary.

cc @vitek-karas @sbomer @tlakollo @mateoatr

Comment (naricc) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml]: @eerhardt I know as of Friday these were necessary to make the test build and run. How late are the "latest"?

Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml]: I guess the support was just for `copy` assemblies. https://github.com/mono/linker/pull/1869. The true fix here is: https://github.com/mono/linker/issues/1837.

Comment (sbomer) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml]: https://github.com/mono/linker/issues/1837 should be fixed by https://github.com/mono/linker/pull/1869/, `copy` or not. If the linker sees a type string that names a forwarder, the forwarder will be kept (and if not it's a bug).

Comment (mateoatr) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml]: > I guess the support was just for copy assemblies.

Yes. The recent changes keep facade assemblies whenever these are referenced from copy assemblies.

Comment (sbomer) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml]: To clarify: that change both:
- keeps facades referenced from metadata in `copy` assemblies (like `typeof(Foo)`, etc.)
- keeps facades referenced by string in _any_ marked code path, `copy` or not

Comment (eerhardt) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml]: @naricc - The absolute latest linker was just merged into main - #51162. Can you try your tests without preserving `System` and `mscorlib` again in these files? If you still need it, we should log a bug so we can fix the linker so this "just works".

Comment (naricc) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml]: I will try that before merging.

Comment (naricc) [reply src/libraries/System.ComponentModel.TypeConverter/tests/ILLink.Descriptor.xml]: @eerhardt It seems they are no longer needed, after rebasing on master, so I have removed them.

**Comment by SamMonoRT** [comment]
@naricc  @radical ---- LGTM - Anything blocking merging this PR ?

**Pull request merged**
